### PR TITLE
SONARIAC-740 S6595: Update cache and install packages in single RUN instruction

### PR DESCRIPTION
**Pull request by petertrr**

### CODE REVIEW

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/PackageManagerUpdateWithoutInstallCheck.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.sonar.iac.docker.checks.utils.CommandDetector;
+import org.sonar.iac.docker.checks.utils.command.SeparatedList;
+import org.sonar.iac.docker.symbols.ArgumentResolution;
+import org.sonar.iac.docker.tree.api.RunInstruction;
+
+@Rule(key = "S6595")
+public class PackageManagerUpdateWithoutInstallCheck implements IacCheck {
+  private static final String MESSAGE = "Update cache and install packages in single RUN instruction.";
+  private static final CommandDetector PACKAGE_MANAGER_DETECTOR = CommandDetector.builder()
+    .with(Set.of("apk", "apt", "apt-get", "aptitude"))
```
Comment (jonas-wielage-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/PackageManagerUpdateWithoutInstallCheck.java:39]: The Set can be extracted into it's own variable as it's used on 2 locations.

File: iac-extensions/docker/src/test/resources/checks/PackageManagerUpdateWithoutInstallCheck/PackageManagerUpdateWithoutInstallCheck.dockerfile [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+apt-get update
+echo APT index updated
+EOF
+
+# Complaiant
+RUN apk update && apk add libcurl
+RUN apt update && apt install libcurl
+RUN apt-get update && apt-get install libcurl
+RUN aptitude update && aptitude install libcurl
+RUN aptitude update && apt-get install libcurl
```
Comment (jonas-wielage-sonarsource) [inline iac-extensions/docker/src/test/resources/checks/PackageManagerUpdateWithoutInstallCheck/PackageManagerUpdateWithoutInstallCheck.dockerfile:37]: Is `aptitude` and `apt-get` using the same cache?
Shouldn't `apt` then use it to? If yes I would add
`RUN apt update && apt-get install libcurl` or some other variations

File: iac-extensions/docker/src/test/resources/checks/PackageManagerUpdateWithoutInstallCheck/PackageManagerUpdateWithoutInstallCheck.dockerfile
Comment (jonas-wielage-sonarsource) [inline iac-extensions/docker/src/test/resources/checks/PackageManagerUpdateWithoutInstallCheck/PackageManagerUpdateWithoutInstallCheck.dockerfile:1]: I would add additional test cases with some other commands in between or test cases where the order of update / install is changed.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/PackageManagerUpdateWithoutInstallCheck.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+  private static void checkPackageManagerInvocations(CheckContext ctx, RunInstruction runInstruction) {
+    SeparatedList<List<ArgumentResolution>, String> splitCommands = ArgumentResolutionSplitter.splitCommands(CheckUtils.resolveInstructionArguments(runInstruction));
+    for (int i = 0; i < splitCommands.elements().size(); i++) {
+      List<ArgumentResolution> argumentResolutions = splitCommands.elements().get(i);
+      for (CommandDetector.Command command : PACKAGE_MANAGER_UPDATE_DETECTOR.search(argumentResolutions)) {
+        List<List<ArgumentResolution>> commandsAfterMatch = splitCommands.elements().subList(i + 1, splitCommands.elements().size());
+        commandsAfterMatch.add(0, argumentResolutions.subList(
+          argumentResolutions.indexOf(command.getResolvedArguments().get(command.getResolvedArguments().size() - 1)) + 1,
+          argumentResolutions.size()));
```
Comment (jonas-wielage-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/PackageManagerUpdateWithoutInstallCheck.java:62]: I can't easily understand this. Why is it added to the list, and what exactly get's added?

File: iac-extensions/docker/src/test/resources/checks/PackageManagerUpdateWithoutInstallCheck/PackageManagerUpdateWithoutInstallCheck.dockerfile (hunk 2 above)
Comment (petertrr) [reply iac-extensions/docker/src/test/resources/checks/PackageManagerUpdateWithoutInstallCheck/PackageManagerUpdateWithoutInstallCheck.dockerfile:37]: Yes, both `apt` and `aptitude` are frontends to `apt-get`. I didn't incorporate any particular checks on whether the same cache is used; I thought that the probability of seeing `apk` and `apt-get` being mixed in a single `RUN` instruction is too low to increase the rule's complexity for it. I'll add some more variations to better illustrate it in tests.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/PackageManagerUpdateWithoutInstallCheck.java (hunk 3 above)
Comment (petertrr) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/PackageManagerUpdateWithoutInstallCheck.java]: Refactored this part a bit

**Review by jonas-wielage-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**